Tags page #738

jquorning opened the pull request:
Generate '_data/tags.json' with list of crates for each tag.
Tags page has section with most popular tags (>= 3 crates) and sections [0-9A-Z] for tags with
that letter. Tags are shown inline with same style as tags on crate pages. On click a list of crates
containing that tag is shown below the tag list. Only one crate list is shown at one time, previous
list is closed. (#738)

jquorning committed:
Add Tags page

Generates '_data/tags.json'. The Tags page is added to top menu. Tags page has
section with most popular tags and section for each tag starting with [A-Z0-9].
Tag links to search page like tags on a crate page. The tag 'ada' is used once
but leads to many search results. Also the two letter tags db, ui, and 7z leads
to search error. 'tags.json' is very terse.

jquorning committed:
Unterse tags.json

jquorning committed:
Add 'popular' variable

jquorning committed:
Tags page to markdown

jquorning committed:
Add unordered lists with crates

filter_ul JS to filter list items in unordered list sections. Mostly HTML
for readability. Permalink.

jquorning committed:
Move tags.json generation

jquorning committed:
Link to tags from crates page only

jquorning committed:
Clearify comments

Fabien-Chouteau commented:
Thank you @jquorning that looks good.

I would like to make the following adjustments to:
-  Make the popular list the top 10 tags by crate count
- Tie the look and feel with the crate page
```diff
diff --git a/crates.md b/crates.md
index 421dd2b..d6db5ed 100644
--- a/crates.md
+++ b/crates.md
@@ -3,9 +3,11 @@ title: Crates
 layout: page
 ---

-[Alire Network Graph](network)
-
-[Tags](/tags/)
+<ul class="nav justify-content-start">
+<li class="nav-item"><a class="nav-link" href="/network/">Network Graph</a></li>
+<li class="nav-item"><a class="nav-link" href="/tags/">Tags</a></li>
+</ul>
+<br>

 {% assign alphabet= "A B C D E F G H I J K L M N O P Q R S T U V W X Y Z" | split: ' ' %}
 {% for letter in alphabet %}
diff --git a/tags.md b/tags.md
index 42ee643..0d351c6 100644
--- a/tags.md
+++ b/tags.md
@@ -3,7 +3,7 @@ title: Tags
 description: "Index of crate tags"
 layout: page
 permalink: /tags/
-popular: 3
+popular_count: 10
 ---

 <script>
@@ -26,7 +26,7 @@ function filter_ul(ul_id, li_id) {
     li = ul.getElementsByTagName("li");
     for (i = 0; i < li.length; i++) {
         if (li[i].id == li_id) {
-            li[i].style.display = "block";
+            li[i].style.display = "list-item";
         } else {
             li[i].style.display = "none";
         }
@@ -42,29 +42,36 @@ function filter_ul(ul_id, li_id) {
 }
 </style>

-## {{ page.title }}
-
-### Popular
-{% capture popular_tags %}
-    {% for tag in site.data.tags %}
-        {% if tag.crates.size >= page.popular %}
-            {{ tag.name }}
-        {% endif %}
-    {% endfor %}
+{% capture tags_with_count %}
+  {% for tag in site.data.tags %}
+    {{ tag.crates.size | plus: 1000000 }}#{{ tag.name }}#{{ tag.crates.size }}
+  {% endfor %}
 {% endcapture %}
-{% assign filtered_pop_list = popular_tags | split: ' ' | sort %}

-{% for tag in filtered_pop_list %}
-<a class="crate-tag-link" href="#popular" onclick="filter_ul('section-Popular','tag-{{ tag }}')">{{ tag }}</a>{% endfor %}
+{% assign sorted_popular_tags = tags_with_count | split:' ' | sort | reverse %}
+{% assign top_popular_tags = sorted_popular_tags | slice: 0, page.popular_count %}
+
+### Top {{ page.popular_count }} Tags
+<div style="white-space: nowrap;">
+{% for tag in top_popular_tags %}
+{% assign tagitems = tag | split: '#' %}
+{% assign name = tagitems[1] %}
+{% assign count = tagitems[2] %}
+<a class="crate-tag-link" onclick="filter_ul('section-Popular','tag-{{ name }}')">{{name}}({{count}})</a>{% endfor %}
+</div>

 <ul id="section-Popular" class="crate_list">
-{% for tag in filtered_pop_list %}
+{% for tag in top_popular_tags %}
+    {% assign tagitems = tag | split: '#' %}
+    {% assign name = tagitems[1] %}
     {%- for crate in site.crates -%}
-        {%- if crate.tags contains tag -%}
-<li id="tag-{{ tag }}"><a href="{{ base_url }}/crates/{{ crate.crate }}">{{ crate.title }}</a> - {{ crate.short_description }}</li>
+        {%- if crate.tags contains name-%}
+<li id="tag-{{ name }}"><a href="{{ base_url }}/crates/{{ crate.crate }}">{{ crate.title }}</a> - {{ crate.short_description }}</li>
 {% endif %}{% endfor %}{% endfor %}
 </ul>

+<br>
+### All Tags
 {% assign alphabet= "0 1 2 3 4 5 6 7 8 9 A B C D E F G H I J K L M N O P Q R S T U V W X Y Z" | split: ' ' %}
 {% for letter in alphabet %}
     {% capture filtered_tags %}
@@ -78,17 +85,15 @@ function filter_ul(ul_id, li_id) {
     {% assign filtered_list = filtered_tags | split: ' ' | sort %}

     {% if filtered_list.size > 0 %}
-
-### {{ letter }}
-
-        {% for tag in filtered_list %}
+<b>{{ letter }}</b>
+{% for tag in filtered_list %}
 <a class="crate-tag-link" href="#{{ letter | downcase }}" onclick="filter_ul('section-{{ letter }}','tag-{{ tag }}')">{{ tag }}</a>{% endfor %}

 <ul id="section-{{ letter }}" class="crate_list">
         {% for tag in filtered_list %}
             {%- for crate in site.crates -%}
                 {%- if crate.tags contains tag -%}
-<li id="tag-{{ tag }}"><a href="{{ base_url }}/crates/{{ crate.crate }}">{{ crate.title }}</a> - {{ crate.short_description }}</li>
+<li id="tag-{{ tag }}"><a href="{{ "crates/" | append: crate.crate | downcase | relative_url }}">{{ crate.title }}</a>: {{ crate.short_description }}</li>
 {% endif %}{% endfor %}{% endfor %}
 </ul>
     {% endif %}
```

Fabien-Chouteau reviewed (changes requested):
See previous message.

jquorning committed:
Top 10 tags

Make the popular list the top 10 tags by crate count
Tie the look and feel with the crate page

Fabien-Chouteau commented:
Thanks @jquorning :+1:
